### DISCUSSION TITLE
Is whole program emulation (external libraries included) possible?

### DISCUSSION THREAD

**Discussion by AminoACID123** [NationalSecurityAgency/ghidra#5445] [Q&A]
How to use ghidra emulator to perform a whole-program emulation?

**Comment by d-millar** [comment]
@AminoACID123 I have to check with @nsadeveloper789 to be sure, but am not aware of a current complete solution to this problem that would make you happy.  There are two issues you are sure to hit.  (1) Even if you've loaded a program into Ghidra with all of its imported libraries, the external references are not resolved by the importer because the various libraries have different address spaces.  So, when you try to step from, say, "entry" to "__libc_start_main", the call will fail because it doesn't really know where to go.  I think this is fixable (and @nsadeveloper789 *may* have a solution), but he's away for a bit.  Problem (2) is the fact that you'll have to model all the SYSCALLs made by the program.  We can't emulate the kernel, so you basically would have to dummy out those calls with the effects you would expect them to have.  Needless to say, this is a lot of work.  Again, @nsadeveloper789 has modellers for some calls, but certainly not the full set.

**Reply by AminoACID123** [reply]
I see. Currently syscalls are not my concerns yet. I am dealing with dynamically-linked executables. Right now I've successfully loaded the main executable and all its dependencies in my ghidrascript, but emulating the whole program would require me to implement some functions of a dynamic linker. I don't know how to do that under the ghidra framework.

**Comment by d-millar** [comment]
Right, that's exactly correct, and I'm not sure the current set of tools support what you want to do.  Again, @nsadeveloper789 would know the answer for sure, but he's on vacation right now.  Will certainly raise the issue when he returns, as I believe what you're describing is a capability many people would want and benefit from.  It's certainly a capability I think we should provide.

**Comment by nsadeveloper789** [comment]
There are two aspects (at least) to this question that I'll try to address here: Loading external modules and implementing system calls. For the first, it's less well understood how to get from where we are in Ghidra now to where we'd like to be to support emulation. Currently, when we import a program (or library) we apply the relocation fixups, but we resolve external addresses to a faked EXTERNAL block. Unfortunately, we have no database table (well, we kind of do, but...) to keep the relocation metadata, so when it's time to load that module into the emulator, I don't have a way to re-fixup the externals to point to their actual addresses in the emulator's memory. The "right" solution is to keep that metadata and properly fixup in the same way the OS would when loading a module. An alternative solution we may explore is to "load" the EXTERNAL block into the emulator and apply some special logic when the PC lands within it. This second option seems to be the shortest path, and not *too bad* a hack, so we're hoping to explore it soon-ish.

For the second (system calls). it's better understood, and we're on our way, but the implementation is vast, so we've left a good bit to the "power user." The new Debugger course materials explain system call implementation in the "Modeling" section. We've done a small bit of file I/O just to flesh out the API and demonstrate it's possible. The rest is left to users, though we *may* wind up providing more out of the box either from contributions or based on demand.

In short, it's not possible out-of-the-box, yet, but we're working on it. If it's few enough modules, and you're into scripting, you might be able to accomplish it now with some work.

**Reply by AminoACID123** [reply]
Thanks for the reply. I have the following questions: 
1. How to keep relocation metadata? One way I can think of is to parse the ELF file again in my ghidra script, but I don't know which API to use.
2. I notice that Ghidra Emulator uses `Memory.setBytes/getBytes` to handle memory access during emulation. Will this incur performance issues when emulating large programs? (Sorry I'm not familiar with the underlying implementation of program memory. I just have this worry.)

**Reply by nsadeveloper789** [reply]
1. I wish this were an easy answer, but unfortunately, Ghidra does not keep sufficient relocation information in an imported image to apply fix-ups when loading multiple modules. This is an issue I've attempted to address internally, but the level of effort required is quite high. Re-parsing the ELF is certainly an option, but it won't generalize. If you're looking to replicate a part of the ELF importer, the place to start is `ElfLoader.java`. The most similar case I can think of to this is in `DbgengX64DisassemblyInject.java`. It re-parses the PE header of Windows modules to determine whether they're 32- or 64-bit. I'm not sure how similar the implementation patterns are, but I suspect you could do similar using `ElfHeader`. Certainly, parsing relocations and fix-ups will be more involved....
2. This depends on which emulator implementation you're using. If you're using the `EmulatorHelper` one (this one backs our p-code tests), I may have to defer to @d-millar. If you're using the `PcodeEmulator` one (this one backs the Debugger/Emulator GUI), then it should only be using `getBytes`. Writes are all cached in memory until it's time to display the state on screen. (the cache reads through to the trace, then the mapped programs). In general, yes, the larger the target memory, the slower things may go. It's probably best to just test and measure yourself. We could speculate, but I think there are too many moving parts to do so accurately. We strive to use hashes or direct indexing as much as possible, but I honestly don't know how things will scale here Big-Oh wise.
